CanvasColorSpaceProposal.md

```diff
@@ -103,61 +84,48 @@ canvas.getContext('2d', { colorSpace: "rec2020",
-* Operations on encoded values always operate on decoded values, not the encoded bits.
-    * I.e. in "unorm8-srgb" encoding, 0xff (1.0) minus 0xbc (0.5) equals 0xbc (0.5).
-
-
-#### Selecting the best color space match for the user agent's display device
-<pre>
-var colorSpace = window.matchMedia("(color-gamut: rec2020)").matches ? "rec2020" :
-    (window.matchMedia("(color-gamut: p3)").matches ? "p3" : "srgb");
-</pre>
+* All input colors (e.g, ``fillStyle`` and ``strokeStyle``) follow the same interpretation as CSS color literals, regardless of the canvas color space.
+* Images drawn to a canvas are converted from their native color space to the backing storage color space of the canvas.
+  * Images with no color profile, when drawn to the canvas, are assumed to be in the sRGB color space.
+* All blending operations are performed in the canvas' backing storage color space.
+  * The midpoint between the backbuffer's colors ``(1,0,0)`` and ``(0,1,0)`` will be ``(0.5,0.5,0)``, even though this is neither the physical nor perceptual midpoint between the two colors.
+  * This means that the same content, rendered with two different canvas backing stores, may be slightly different.
+* The functions ``toDataURL`` and ``toBlob`` should produce resources that best match the fidelity of the underlying canvas.
+  * Subject to the limitations of the implementation and the requested format.
 
-#### Selecting the best encoding for the user agent's display device
-Selection should be based on the best color space match (see above). For srgb, at least 8 bits per component is recommended; for p3, 10 bits; and for rec2020, 12 bits.  The float16 format is suitable for any colorspace.  There may soon be a proposal to add a way of detecting HDR displays, possibly something like "window.screen.isHDR()" (TBD), which would be a good hint to use the float16 format.
+#### WebGL behavior
 
-#### Non-standard color spaces
-For future consideration: support could be added for color space defined using the [CSS @color-profile rule](https://www.w3.org/TR/css-color-4/#at-profile).
+* Values stored in WebGL backbuffers are in the canvas's color space.
+* Values written by ``gl_FragColor`` use the primaries of the canvas' color space.
+* For the color encoding of ``"unorm-srgb8"``, the interpretation of values written by ``gl_FragColor`` is as follows:
+  * The color value is in a linear color space.
+  * The inverse sRGB transfer function is applied only after blending has occurred.
+  * E.g, if the value ``0.5`` is written in ``gl_FragColor``, the result will be stored as ``0xbc``.
+* For all other color encodings, the value written by ``gl_FragColor`` is in the canvas' color space.
```

Comment:
This is sort of subtly wrong, or maybe just misleading: For unorm8-srgb (as implemented in hardware), 0xbc *is* the encoding of 0.5, just as unorm8's is 0x80 and float32's is 0x3f000000.
There's more ways than just gl_FragColor to set color data in WebGL, and leaning heavily on "encoding" as our word of choice communicates this with proper connotation.

I think having a description of the unusual unorm8-srgb encoding is essential, it would help to have it next to a description of the widely-understood unorm8 encoding. Having these in sequence would be helpful by drawing important parallels between the two similar-but-different encodings, jump-starting strong intuition.

```
function encodeUnorm8(val) { return val / 0xff; }
function encodeUnorm8Srgb(val) {
  if (val < 0.0031308) return 12.92 * val;
  return 1.055 * Math.pow(val, 0.41666) - 0.055;
}
```
(from https://www.khronos.org/registry/OpenGL/specs/es/3.0/es_spec_3.0.pdf p182)

Reply:
Agree that leaning on the idea of (HW) sRGB being an encoding is the better mental model. Also, I'm guessing you meant to say that encodeUnorm8 should be { return val ***** 0xff; }

Reply:
I think that support for "uint8-srgb" should be a separate WebGL spec that, in effect, makes the backbuffer be SRGB8_ALPHA8 (or similar).

I am of the opinion that it does not belong in this specification -- it doesn't have a clear meaning for 2D canvas, it's independent of color space, and it is only an option for unorm8 canvases.

Reply:
We should probably have a larger discussion about this. I think there is desire unorm8-srgb canvas2d (@kenrussell?), but there are real implementability/feasibility concerns. Can you file an issue?

Reply:
I don't remember off the top of my head - might have been discussed a while ago with @fserb and others. For 2D canvas photo editing applications, I can imagine it might be desirable to use the unorm8-srgb encoding to get more detail in darker regions of images.

Could we please leave the unorm8-srgb description in the WebGL-specific section of this document for the time being? Chris, would you be willing to add Jeff's code snippets?

Reply:
The actual bits-stored-in-the-framebuffer have identical meanings when using uint8-srgb and uint8. They are both in the sRGB color space.

What is different is that when a uint8-srgb texture is sampled in WebGL, the values are converted into a linear-sRGB color space by the texturing unit. Similarly, when values are written to the framebuffer, they are converted from linear-sRGB color space down to sRGB color space before being written (but after blending).

It's a really texture-unit-and-framebuffer-specific thing, which is why it doesn't clear meaning in other contexts (e.g, 2D canvas). A few years ago we considered exposing a 2D canvas that was sort-of-in-linear-space-but-stored-as-sRGB, and it was an unintuitive mess (and was ultimately one of the things that caused this spec's implementation to stall out as it did).

Reply:
For the next incremental step of this specification, @jdashg and I are in agreement that we'd like the unorm8-srgb option to remain. The discussion of applicability to 2D canvas can be deferred, and would perhaps be better done on a conference call, since there's disagreement about its utility.

Can we please resolve this particular conversation? May I suggest incorporating this text:

- The encodings for a specific color value differ between <code>unorm8</code> and <code>unorm8-srgb</code> as follows:

<pre>
function encodeUnorm8(val) { return val * 0xff; }
function encodeUnorm8Srgb(val) {
  if (val < 0.0031308) return 12.92 * val;
  return 1.055 * Math.pow(val, 0.41666) - 0.055;
}
</pre>

Reply:
Oops encodeUnorm8Srgb needs to include the `* 0xff` on its returns, sorry!

Reply:
I've updated the text to include kbr's suggestion about encoding.

I'd like to continue iterating on whether GL color encoding should be specified independently of the storage format, but we can defer that discussion to some other time.

Reply:
Thanks @ccameron-chromium . Let me call this conversation resolved, and we can open a new one against the most recent version of this pull request if necessary.